src/UserSpaceInstrumentation/Trampoline.cpp

```diff
@@ -161,4 +176,124 @@ ErrorMessageOr<uint64_t> AllocateMemoryForTrampolines(pid_t pid, const AddressRa
   return AllocateInTracee(pid, address_range.start, size);
 }
 
+ErrorMessageOr<int32_t> AddressDifferenceAsInt32(uint64_t a, uint64_t b) {
+  const uint64_t abs_diff = (a > b) ? (a - b) : (b - a);
+  constexpr uint64_t kAbsMaxInt32AsUint64 =
+      static_cast<uint64_t>(std::numeric_limits<int32_t>::max());
+  constexpr uint64_t kAbsMinInt32AsUint64 =
+      static_cast<uint64_t>(-static_cast<int64_t>(std::numeric_limits<int32_t>::min()));
+  if ((a > b && abs_diff > kAbsMaxInt32AsUint64) || (b > a && abs_diff > kAbsMinInt32AsUint64)) {
+    return ErrorMessage("Difference is larger than +-2GB.");
+  }
+  return a - b;
+}
+
+ErrorMessageOr<RelocatedInstruction> RelocateInstruction(cs_insn* instruction, uint64_t old_address,
+                                                         uint64_t new_address) {
+  RelocatedInstruction result;
+  if ((instruction->detail->x86.modrm & 0xC7) == 0x05) {
+    // The modrm byte can encode a memory operand as a signed 32 bit displacement on the rip. See
+    // "Intel 64 and IA-32 Architectures Software Developer’s Manual Vol. 2A" Chapter 2.1.
+    // Specifically table 2-2.
+    const int32_t old_displacement = *absl::bit_cast<int32_t*>(
+        instruction->bytes + instruction->detail->x86.encoding.disp_offset);
+    const uint64_t old_absolute_address = old_address + instruction->size + old_displacement;
+    ErrorMessageOr<int32_t> new_displacement_or_error =
+        AddressDifferenceAsInt32(old_absolute_address, new_address + instruction->size);
+    if (new_displacement_or_error.has_error()) {
+      return ErrorMessage(absl::StrFormat(
+          "While trying to relocate an instruction with rip relative addressing the target was out "
+          "of range from the trampoline. old address: %#x, new address :%#x instruction: %s",
+          old_address, new_address, InstructionBytesAsString(instruction)));
+    }
+    result.code.resize(instruction->size);
+    memcpy(result.code.data(), instruction->bytes, instruction->size);
+    *absl::bit_cast<int32_t*>(result.code.data() + instruction->detail->x86.encoding.disp_offset) =
+        new_displacement_or_error.value();
+  } else if (instruction->detail->x86.opcode[0] == 0xeb ||
+             instruction->detail->x86.opcode[0] == 0xe9) {
```

Comment:
For all the branches, I would suggest also showing the assembly of the instruction (or one example if the branch covers multiple opcodes) before the relocation, in addition to the description. I think it helps understanding the different cases.

Reply:
I added examples of original code .... please take a look if this solves the problem.

Reply:
I was thinking of the same also for the branches below, or do you think it doesn't really make sense/it's already enough in the tests?